File: blind75_grind75/LongestSubstringWithoutRepeatingCharacters.java
package grind75;

import java.util.Arrays;
import java.util.HashSet;

public class LongestSubstringWithoutRepeatingCharacters {

    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1) return s.length();
        int n = s.length(), ans = 0;
        int[] index = new int[128];
        // try to extend the range [i, j]
        for (int j = 0, i = 0; j < n; j++) {
            // System.out.println("i===" + i);
            i = Math.max(index[s.charAt(j)], i);
            // System.out.println("i===" + i);
            ans = Math.max(ans, j - i + 1);
            // System.out.println("s.charAt(j):" + s.charAt(j));
            // System.out.println("index[s.charAt(j)]:" + index[s.charAt(j)]);
            // System.out.println(Arrays.toString(index));
            index[s.charAt(j)] = j + 1;
            // System.out.println(Arrays.toString(index));
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println(new LongestSubstringWithoutRepeatingCharacters().lengthOfLongestSubstring("anviaj"));
    }
}

File: go/grind75/LongestSubstringWithoutRepeatingCharacters.go
package grind75

func lengthOfLongestSubstring(s string) int {
	if len(s) < 2 {
		return len(s)
	}
	maxlen := 0
	table := map[byte]int{}
	for i, j := 0, 0; j < len(s); j++ {
		c := s[j]

		if table[c] != 0 && table[c] >= i {
			i = table[c] + 1
		}
		maxlen = max(maxlen, j-i+1)
		table[c] = j
	}
	return maxlen
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
